### PR TITLE
Expose selectionHeightStyle and SelectionWidthStyle on SelectableText

### LINKED ISSUE #67434: Text background color height is inconsistent when english/non-english characters are used together
## Steps to Reproduce
This is reproducible in all places where a background color can be added to text(i.e. `style` within `Text` or `TextSpan`, or in my case, `SelectableText`/`Text`)
```
class DemoText extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _DemoTextState();
  }
}

class _DemoTextState extends State<DemoText> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Builder(
        builder: (BuildContext buildContext) {
          return Scaffold(
              body: Center(
                  child: Text(
            'test测试test',
            style: Theme.of(context).textTheme.bodyText2.copyWith(
                  color: Colors.black,
                  background: Paint()..color = Colors.orange,
                ),
          )));
        },
      ),
    );
  }
}
```

Similar for 
```
Scaffold(
            body: Center(child: SelectableText('test测试test', ))
)
```

**Expected results:** 
 Height for all background color should  be the same

**Actual results:** 

demo

Non-english characters is higher(here it is "测试"),which makes it looks weird
<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel stable, 1.22.0, on Mac OS X 10.15.6 19G2021, locale en-US)
    • Flutter version 1.22.0 at /Users/edward/bin/flutter
    • Framework revision d408d302e2 (7 days ago), 2020-09-29 11:49:17 -0700
    • Engine revision 5babba6c4d
    • Dart version 2.10.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/edward/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.8.3

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin installed
    • Dart plugin version 202.7319.5

[!] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (1 available)
    • iPhone (mobile)  iOS 14.0.1

! Doctor found issues in 2 categories.
```

</details>

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
## Description
This brings over #48917 to `SelectableText`. Currently Flutter's `SelectableText` only allows us to achieve behavior 1 from the screenshots below. With this change developers are now able to achieve behavior 2 which is more inline with the behavior on native android, iOS and the web. 

This is achieved through exposing the `SelectionHeightStyle` and `SelectionWidthStyle` properties on `SelectableText` (before this change they always default to a `tight` box style), which are passed to `GetBoxesForRange` to allow for different kinds of boxes for selection. 

1|2
---|---
Simulator Screen Shot - iPhone 12 Pro Max - 2021-06-28 at 10 11 17|Simulator Screen Shot - iPhone 12 Pro Max - 2021-06-28 at 10 23 14

## Related Issues
Closes #67434

## Tests
I added the following tests:
* SelectableText
  * Golden tests to verify correct selection height styles.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/85640

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/85640

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/85640

File: packages/flutter/test/widgets/selectable_text_test.dart [hunk 1]
```diff
@@ -4799,4 +4801,102 @@ void main() {
...
     expect(selection!.extentOffset, 14);
   }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS, TargetPlatform.android }));
+
+  testWidgets('text selection style 1', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Material(
+          child: Center(
+            child: Column(
+              children: <Widget>[
```
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/test/widgets/selectable_text_test.dart:4811]: Same for below on the second test.

**Review by justinmc — APPROVED** [review]
LGTM besides a quick code style thing in the tests.

Looks like you've got goldens failures, is it the ones you added that are failing?

CC @GaryQian since this is related to your https://github.com/flutter/flutter/pull/48917.

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/85640

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/85640

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/85640

File: packages/flutter/test/widgets/selectable_text_test.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (Renzo-Olivares) [reply packages/flutter/test/widgets/selectable_text_test.dart]: Which widget are you referring to?

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/test/widgets/selectable_text_test.dart]: Oh sorry, you're good.  I was misreading the open/close brackets for the parameters in SelectableText.rich.

**Review by GaryQian — APPROVED** [review]
These changes LGTM as well, just double check to tests 🙂

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 11.
View them at https://flutter-gold.skia.org/cl/github/85640

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 12.
View them at https://flutter-gold.skia.org/cl/github/85640

**Pull request merged**